### DISCUSSION TITLE
private traits?

### DISCUSSION THREAD

**Discussion by eadf** [bitshifter/glam-rs#241] [Q&A] [answered]
I'm trying to implement a type containing a `XYZ<T> `where `T` is intended to be any builtin float or int:
```
pub struct A<T> {
    pub v: glam::XYZ<T>,
}

impl<T:glam::core::traits::scalar::NumEx> A<T> {
    fn fuubar(&self) -> () {
        self.v + self.v;
    }
}

/*
impl A<i32> {
    fn fuubar(&self) -> () {
        self.v + self.v;
    }
}
*/
```

Obviously this does not work as `glam::core::traits::scalar::NumEx` is private. The second impl fails too, because it does not know how to add two `XYZ<i32>` (can't import the `glam::core::traits::vector::Vector` trait)

Even something simple as `glam::XYZ::<i32>::new(1,2,3);` refuses to compile (glam::core::traits::vector::Vector3 is private).

It feels like I'm overlooking something really simple here, but I can't find it

**Comment by bitshifter** [comment]
`XYZ` is only intended for dereferencing the `.x`, `.y` and `.z` components and has no other functionality of it's own. The private implementations are internal details and aren't intended for external use. What are you trying to achieve?

**Reply by eadf** [reply]
A generic type that contains boundary indications, like an AABB, but it needs to be able to contain ints _or_ floats.

**Comment by bitshifter** [comment, marked as answer]
Ultimately glam is not a generic library. Some generics are used internally but the internal implementation is subject to change.

I have a work in progress architecture document that goes into more depth here https://github.com/bitshifter/glam-rs/blob/architecture/ARCHITECTURE.md

**Reply by eadf** [reply]
Ok, thanks for your patience. I know I should have spent more time rtfm:ing before posting this.

**Comment by bitshifter** [comment]
It's really no problem, I'm still trying to write the manual 🙂

The deref to XYZ does confuse people as well.
